packages/pymedphys_dicom/src/pymedphys_dicom/ct/extend.py

```diff
@@ -0,0 +1,180 @@
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License and the additional terms for more
+# details.
+
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+# ADDITIONAL TERMS are also included as allowed by Section 7 of the GNU
+# Affero General Public License. These additional terms are Sections 1, 5,
+# 6, 7, 8, and 9 from the Apache License, Version 2.0 (the "Apache-2.0")
+# where all references to the definition "License" are instead defined to
+# mean the AGPL-3.0+.
+
+# You should have received a copy of the Apache-2.0 along with this
+# program. If not, see <http://www.apache.org/licenses/LICENSE-2.0>.
+
+
+import os
+from copy import deepcopy
+from glob import glob
+import datetime
+import random
+from collections import deque
+
+import pydicom
+from ..constants.uuid import PYMEDPHYS_ROOT_UID
+
+
+def extend(input_dir, output_dir, index_to_copy, number_of_slices):
+    filepaths = glob(os.path.join(input_dir, '*'))
+    dicom_datasets = load_dicom_into_deque(filepaths)
+    extend_datasets(dicom_datasets, index_to_copy, number_of_slices)
+
+    filenames = [
+        os.path.basename(filepath)
+        for filepath in filepaths
+    ]
+
+    common_prefix = os.path.commonprefix(filenames)
+
+    number_of_digits = len(str(len(dicom_datasets)))
+    new_filenames = [
+        "{}{}.dcm".format(
+            common_prefix,
+            str(dicom_dataset.InstanceNumber).zfill(number_of_digits)
+        )
+        for dicom_dataset in dicom_datasets
+    ]
+
+    new_filepaths = [
+        os.path.join(output_dir, filename)
+        for filename in new_filenames
+    ]
+
+    for dicom_dataset, filepath in zip(dicom_datasets, new_filepaths):
+        dicom_dataset.save_as(filepath)
+
+
+def extend_datasets(dicom_datasets, index_to_copy, number_of_slices, uids=None):
+    copy_slices_and_append(dicom_datasets, index_to_copy, number_of_slices)
+    refresh_instance_numbers(dicom_datasets)
+    generate_new_uids(dicom_datasets, uids=uids)
+
+
+def load_dicom_into_deque(filepaths):
+    dicom_datasets_initial_read = [
+        pydicom.dcmread(filepath, force=True)
+        for filepath in filepaths
+    ]
+
+    dicom_datasets = convert_datasets_to_deque(dicom_datasets_initial_read)
+
+    return dicom_datasets
+
+
+def convert_datasets_to_deque(datasets):
+    dicom_datasets = deque()
+
+    for dicom_dataset in sorted(datasets, key=slice_location):
+        dicom_datasets.append(dicom_dataset)
+
+    return dicom_datasets
+
+
+
+def instance_number(dicom_dataset):
+    return dicom_dataset.InstanceNumber
+
+
+def slice_location(dicom_dataset):
+    return float(dicom_dataset.SliceLocation)
+
+
+def copy_slices_and_append(dicom_datasets, index_to_copy, number_of_slices):
```

Comment:
Adds to confusion. Should `dicom_datasets` be a `deque`?

Reply:
Key is, it doesn't have to be a deque. As long as the provided append method exists on the object.

Reply:
If it quacks like a duck, it can go in there:

https://dev.to/alvarocavalcanti/practical-python---duck-typing-to-the-rescue-46e6

Reply:
(A list will work for appending slices to the end)

Reply:
Fair enough for this specific function, but does that apply to all of your `deque` functions?

Reply:
If so, ignore all my comments regarding renaming `datasets` (except perhaps the still arguably superfluous `dicom` hehe). If all uses of `deque` are truly perfectly compatible with `list`, I'm happy :)

Reply:
A deque is essentially a list, just stored in memory in a way optimised for appending and removing items from either side...

Reply:
List doesn't have an 'appendleft' method, but that will raise an error if one tries to append left with a list.

Reply:
Fair enough. :)